### PR TITLE
[Merged by Bors] - feat(Logic): Godel Beta Function (toward first incompleteness theorem)

### PR DESCRIPTION
**Pull request by hmonroe**
This file proves Gödel's Beta Function Lemma, used to prove the First Incompleteness Theorem. This code is a step towards eventually including a proof of Gödel's First Incompleteness Theorem
and other key results from the repository https://github.com/iehality/lean4-logic.

---

- [x] depends on: #9005

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Logic/Godel/GodelBetaFunction.lean [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+* [R. Kaye, *Models of Peano arithmetic*][kaye1991]
+* <https://en.wikipedia.org/wiki/G%C3%B6del%27s_%CE%B2_function>
+
+## Tags
+
+Gödel, beta function
+-/
+
+namespace Nat
```
Comment (tb65536) [inline Mathlib/Logic/Godel/GodelBetaFunction.lean:39]: These first few nat lemmas that aren't specific to the GodelBetaFunction should be moved to earlier files in the library.

File: Mathlib/Logic/Godel/GodelBetaFunction.lean [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+/-- Gödel's Beta Function -/
+def beta (n i : ℕ) := n.unpair.1 % ((i + 1) * n.unpair.2 + 1)
+
+/-- Inverse of Gödel's Beta Function -/
+def unbeta (l : List ℕ) :=
+  (chineseRemainderList (coprimeList l) (coprimes_coprimeList l) : ℕ).pair (listSup l)!
+
+/-- **Gödel's Beta Function Lemma** -/
+lemma beta_function_lemma (l : List ℕ) (i : Fin l.length) :
```
Comment (fpvandoorn) [inline Mathlib/Logic/Godel/GodelBetaFunction.lean:167]: We already have [List.encodable](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Logic/Equiv/List.html#List.encodable). How is this different from `encode`/`decode` for lists?

Comment (fpvandoorn) [reply Mathlib/Logic/Godel/GodelBetaFunction.lean:167]: Or perhaps even more closely related: [Equiv.listNatEquivNat](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Logic/Equiv/List.html#Equiv.listNatEquivNat)

Comment (iehality) [reply Mathlib/Logic/Godel/GodelBetaFunction.lean:167]: Goedel's beta function is used to prove that arithmetically definable functions are equal to `Partrec'`. ([proof](https://github.com/iehality/lean4-logic/blob/7a12751321928ed48c565b6db906d1d8a303ed25/Logic/Vorspiel/Arith.lean#L492)) While `List.encodable` is computable, it is not immediately apparent whether it is arithmetically definable.

Comment (fpvandoorn) [reply Mathlib/Logic/Godel/GodelBetaFunction.lean:167]: But it is arithmetically definable. `encodeList` and `decodeList` are defined in terms of `Nat.pair` and `Nat.unpair`, and these are arithmetically definable (I guess it might be hardest to prove that for `Nat.sqrt`). 
I think it would be nicer to prove this reusing existing Mathlib definitions.

Comment (iehality) [reply Mathlib/Logic/Godel/GodelBetaFunction.lean:167]: The term arithmetically definable function, we mean here a class of functions whose graph can be described by a $\Sigma_1$ formula of arithmetic.  While it is straightforward to prove that this class is closed under function composition, but whether it is closed in primitive recursion is a proposition demanding a technical proof (we use beta function lemma to prove this). Since `encodeList` and `decodeList` are defined using recursion, they cannot be used to prove this.

File: Mathlib/Logic/Godel/GodelBetaFunction.lean [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+  simpa[show a % n = b % n from h] using mod_eq_of_lt hb
+
+lemma coprime_list_prod_iff_right {k} {l : List ℕ} :
+    Coprime k l.prod ↔ ∀ n ∈ l, Coprime k n := by
+  induction' l with m l ih <;> simp[Nat.coprime_mul_iff_right, *]
+
+/-- Define a list of coprimes -/
+inductive Coprimes : List ℕ → Prop
+  | nil : Coprimes []
+  | cons {n : ℕ} {l : List ℕ} : (∀ m ∈ l, Coprime n m) → Coprimes l → Coprimes (n :: l)
```
Comment (eric-wieser) [inline Mathlib/Logic/Godel/GodelBetaFunction.lean:52]: Is this `List.Pairwise Coprime`?

File: docs/references.bib [hunk 4]
```diff
@@ -8,19 +8,6 @@
...
-                  {Semantic} structures},
-  isbn          = {0-19-853762-X},
-  year          = {1994},
-  publisher     = {Oxford: Clarendon Press},
-  language      = {English},
-  keywords      = {68-00,68Q55},
-  zbmath        = {783754},
-  zbl           = {0829.68111}
-}
-
```
Comment (eric-wieser) [inline docs/references.bib:23]: Why did you delete these?

Comment (hmonroe) [reply docs/references.bib]: Not sure how but I have fixed it

File: Mathlib/Data/ZMod/Coprime.lean [hunk 5]
```diff
@@ -37,4 +37,112 @@ theorem eq_zero_of_gcd_ne_one {a : ℤ} {p : ℕ} (pp : p.Prime) (h : a.gcd p 
...
+    exact pp
+    )
+
+lemma pairwise_coprime_coprimeList (l : List ℕ) : ((coprimeList l).map Prod.snd).Pairwise
+    Coprime := by
+  have : (coprimeList l).map Prod.snd = List.ofFn (fun i : Fin l.length => (i + 1) *
+      (listSup l)! + 1) := by
+    simp[coprimeList, Function.comp]
+  rw[this]
+  exact pairwise_coprime_of_nodup
```
Comment (eric-wieser) [inline Mathlib/Data/ZMod/Coprime.lean:127]:
```suggestion
  exact List.Nodup.pairwise_of_forall_ne
```

File: Mathlib/Logic/Godel/GodelBetaFunction.lean (hunk 1 above)
Comment (hmonroe) [reply Mathlib/Logic/Godel/GodelBetaFunction.lean:40]: I have moved them to Mathlib.Data.Zmod.Coprime

File: Mathlib/Logic/Godel/GodelBetaFunction.lean (hunk 3 above)
Comment (hmonroe) [reply Mathlib/Logic/Godel/GodelBetaFunction.lean]: Comment taken

File: Mathlib/Data/ZMod/Coprime.lean [hunk 6]
```diff
@@ -37,4 +37,118 @@ theorem eq_zero_of_gcd_ne_one {a : ℤ} {p : ℕ} (pp : p.Prime) (h : a.gcd p 
   (@eq_zero_iff_gcd_ne_one a p ⟨pp⟩).mpr h
 #align zmod.eq_zero_of_gcd_ne_one ZMod.eq_zero_of_gcd_ne_one
 
+open Nat
+
+lemma coprime_list_prod_iff_right {k} {l : List ℕ} :
+    Coprime k l.prod ↔ ∀ n ∈ l, Coprime k n := by
+  induction' l with m l ih <;> simp[Nat.coprime_mul_iff_right, *]
+
+lemma pairwise_coprime_of_nodup {l : List ℕ} (hl : l.Nodup) (H : ∀ n ∈ l, ∀ m ∈ l, n ≠ m →
```
[the remark this answers is not in the document]
Comment (hmonroe) [reply Mathlib/Data/ZMod/Coprime.lean:46]: Yes I have fixed that

File: Mathlib/Data/ZMod/Coprime.lean [hunk 7]
```diff
@@ -37,4 +37,118 @@ theorem eq_zero_of_gcd_ne_one {a : ℤ} {p : ℕ} (pp : p.Prime) (h : a.gcd p 
   (@eq_zero_iff_gcd_ne_one a p ⟨pp⟩).mpr h
 #align zmod.eq_zero_of_gcd_ne_one ZMod.eq_zero_of_gcd_ne_one
 
+open Nat
+
+lemma coprime_list_prod_iff_right {k} {l : List ℕ} :
+    Coprime k l.prod ↔ ∀ n ∈ l, Coprime k n := by
```
[the remark this answers is not in the document]
Comment (eric-wieser) [reply Mathlib/Data/ZMod/Coprime.lean:43]: In fact, this would make a good standalone PR; along with the `Multiset.prod` and `Finset.prod` versions.

File: Mathlib/Logic/Godel/GodelBetaFunction.lean (hunk 2 above)
Comment (hmonroe) [reply Mathlib/Logic/Godel/GodelBetaFunction.lean:59]: OK I have added that to the documentation. Thanks Floris.

File: Mathlib/Logic/Godel/GodelBetaFunction.lean [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+Gödel, beta function
+-/
+
+namespace Nat
+
+open ZMod
+
+/-- Gödel's Beta Function. This is similar to List.encodeList, but is arithmetically definable. -/
+def beta (n i : ℕ) := n.unpair.1 % ((i + 1) * n.unpair.2 + 1)
```
[the remark this answers is not in the document]
Comment (hmonroe) [reply Mathlib/Logic/Godel/GodelBetaFunction.lean]: OK I added the return type

File: Mathlib/Logic/Godel/GodelBetaFunction.lean [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+## Tags
+
+Gödel, beta function
+-/
+
+namespace Nat
+
+open ZMod
+
+/-- Gödel's Beta Function. This is similar to `Encodable.decodeList`, but it is easier to prove that
```
[the remark this answers is not in the document]
Comment (hmonroe) [reply Mathlib/Logic/Godel/GodelBetaFunction.lean]: ok

File: Mathlib/Logic/Godel/GodelBetaFunction.lean [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+it is arithmetically definable. -/
+def beta (n i : ℕ) : ℕ := n.unpair.1 % ((i + 1) * n.unpair.2 + 1)
+
+/-- Inverse of Gödel's Beta Function. This is similar to `Encodable.encodeList` , but it is easier
```
[the remark this answers is not in the document]
Comment (hmonroe) [reply Mathlib/Logic/Godel/GodelBetaFunction.lean]: Does this PR need more work or is it ready? Many thanks

File: Mathlib/Data/ZMod/Coprime.lean [hunk 11]
```diff
@@ -37,4 +37,114 @@ theorem eq_zero_of_gcd_ne_one {a : ℤ} {p : ℕ} (pp : p.Prime) (h : a.gcd p 
   (@eq_zero_iff_gcd_ne_one a p ⟨pp⟩).mpr h
 #align zmod.eq_zero_of_gcd_ne_one ZMod.eq_zero_of_gcd_ne_one
 
+open Nat
```
Comment (eric-wieser) [inline Mathlib/Data/ZMod/Coprime.lean:40]: All of the results below this line end up in the `ZMod` namespace, but none of them are about `ZMod`! I think they belong somewhere else.

File: Mathlib/Data/ZMod/Coprime.lean [hunk 12]
```diff
@@ -37,4 +37,114 @@ theorem eq_zero_of_gcd_ne_one {a : ℤ} {p : ℕ} (pp : p.Prime) (h : a.gcd p 
...
+    constructor
+    · intro h; exact ⟨by simpa using h ⟨0, by simp⟩, fun i => by simpa using h i.succ⟩
+    · intro h i
+      cases i using Fin.cases
+      · simpa using h.1
+      · simpa using h.2 _
+
+/-- List of coprimes, using the Chinese remainder theorem -/
+def chineseRemainderList : (l : List (ℕ × ℕ)) → (H : (l.map Prod.snd).Pairwise Coprime) →
+    { k // ∀ i, k ≡ (l.get i).1 [MOD (l.get i).2] }
```
Comment (eric-wieser) [inline Mathlib/Data/ZMod/Coprime.lean:72]: Please explain in the docstring what `chineseRemainderList l hl` means; your current description doesn't mention `l` at all.

Comment (eric-wieser) [reply Mathlib/Data/ZMod/Coprime.lean:72]: Further more, this name is bizarre because the return value isn't a list at all, so this certainly is not a "list of coprimes"!

File: Mathlib/Data/ZMod/Coprime.lean (hunk 7 above)
[the remark this answers is not in the document]
Comment (eric-wieser) [reply Mathlib/Data/ZMod/Coprime.lean:43]: I'd encourage you to do that PR first; this lemma still looks like its in the wrong place, and that PR would probably be quicker to go through.

File: Mathlib/Data/ZMod/Coprime.lean (hunk 11 above)
Comment (iehality) [reply Mathlib/Data/ZMod/Coprime.lean]: Moved theorem related to `chineseRemainderList` (remaned to `chineseRemainderOfList`) to Data.Nat.ModEq.

File: Mathlib/Data/ZMod/Coprime.lean (hunk 12 above)
Comment (iehality) [reply Mathlib/Data/ZMod/Coprime.lean]: Renamed to `chineseRemainderOfList`. Also added an explanation to make it clear that it is a list version of `chineseRemainder`.

File: Mathlib/Data/ZMod/Coprime.lean (hunk 7 above)
[the remark this answers is not in the document]
Comment (hmonroe) [reply Mathlib/Data/ZMod/Coprime.lean]: While PR 9005 is pending, should I amend this PR, for instance, remove "awaiting-review", make it a branch of 9005, close, do nothing?

[the remark this answers is not in the document]
Comment (eric-wieser) [reply Mathlib/Data/ZMod/Coprime.lean]: You should write `#9005` in the PR description in the "dependent PRs" section, as described in the PR template.

[the remark this answers is not in the document]
Comment (hmonroe) [reply Mathlib/Data/ZMod/Coprime.lean]: I just did that. I suppose I can wait for #9005 to go through before revising this one to be consistent with that, right?

**Comment by eric-wieser** [comment]
Please run `git merge origin/master` to shrink the diff on this PR. It looks like you just cherry-picked/rebased your other PR, which results it github showing it as not already merged.

**Comment by hmonroe** [comment]
> Please run `git merge origin/master` to shrink the diff on this PR. It looks like you just cherry-picked/rebased your other PR, which results it github showing it as not already merged.

Did we fix the issue

**Comment by hmonroe** [comment]
What is the next step for this PR--is any further action needed?

File: Mathlib/Data/Nat/ModEq.lean [hunk 13]
```diff
@@ -530,4 +543,129 @@ theorem odd_mod_four_iff {n : ℕ} : n % 2 = 1 ↔ n % 4 = 1 ∨ n % 4 = 3 :=
...
+  · simp [modEq_one]
+  · have : Coprime (s i) (l.map s).prod := coprime_list_prod_right_iff.mpr
+      (by simp; intro j hj; exact (List.pairwise_cons.mp co).1 j hj)
+    simp [← modEq_and_modEq_iff_modEq_mul this, ih (List.Pairwise.of_cons co)]
+
+variable (a s : ι → ℕ)
+
+/-- The natural number less than `(l.map s).prod` congruent to
+`a i` mod `s i` for all  `i ∈ l`. -/
+def chineseRemainderOfList : (l : List ι) → l.Pairwise (Coprime on s) →
```
Comment (awainverse) [inline Mathlib/Data/Nat/ModEq.lean:575]: Why is the return type `{ k // ∀ i ∈ l, k ≡ a i [MOD s i] }`?
I would guess that this is easier to work with as just returning `k : ℕ`, and then having a separate proof, called `chineseRemainderofList_spec` or something, that when `k` is the output, `∀ i ∈ l, k ≡ a i [MOD s i]`.

File: Mathlib/Data/Nat/ModEq.lean [hunk 14]
```diff
@@ -530,4 +543,129 @@ theorem odd_mod_four_iff {n : ℕ} : n % 2 = 1 ↔ n % 4 = 1 ∨ n % 4 = 3 :=
...
+      (chineseRemainderOfList_lt_prod _ _ _ _ hs)
+      (by rw [hlp]
+          exact chineseRemainderOfList_lt_prod _ _ _ _
+            (by simpa [List.Perm.mem_iff hl.symm] using hs))
+
+/-- The natural number less than `(m.map s).prod` congruent to
+`a i` mod `s i` for all  `i ∈ m`. -/
+def chineseRemainderOfMultiset {m : Multiset ι} :
+    m.Nodup → (∀ i ∈ m, s i ≠ 0) → Set.Pairwise {x | x ∈ m} (Coprime on s) →
+    { k // ∀ i ∈ m, k ≡ a i [MOD s i] } :=
```
Comment (awainverse) [inline Mathlib/Data/Nat/ModEq.lean:629]: Again, would it be better to separate out the data that's the output from the proof it works?

File: Mathlib/Data/Nat/ModEq.lean [hunk 15]
```diff
@@ -530,4 +543,129 @@ theorem odd_mod_four_iff {n : ℕ} : n % 2 = 1 ↔ n % 4 = 1 ∨ n % 4 = 3 :=
...
+    chineseRemainderOfMultiset a s nod hs pp < (m.map s).prod := by
+  induction' m using Quot.ind with l
+  unfold chineseRemainderOfMultiset; simp
+  exact chineseRemainderOfList_lt_prod a s l
+    (List.Nodup.pairwise_of_forall_ne nod pp) (by simpa using hs)
+
+/-- The natural number less than `∏ i in t, s i` congruent to
+`a i` mod `s i` for all  `i ∈ t`. -/
+def chineseRemainderOfFinset (t : Finset ι)
+    (hs : ∀ i ∈ t, s i ≠ 0) (pp : Set.Pairwise t (Coprime on s)) :
```
Comment (awainverse) [inline Mathlib/Data/Nat/ModEq.lean:661]: Again, would it be better to separate out the data that's the output from the proof it works?

**Review by awainverse — COMMENTED** [review]
This looks good to me. Like Floris, I have some reservations as to whether the best way to do this is with the traditional beta function, but it has been used well in the project this is being ported from, so it seems fine.

File: Mathlib/Data/Nat/ModEq.lean (hunk 13 above)
Comment (iehality) [reply Mathlib/Data/Nat/ModEq.lean:575]: I defined like this because I followed the definition of [Nat.chineseRemainder](https://leanprover-community.github.io/mathlib4_docs//Mathlib/Data/Nat/ModEq.html#Nat.chineseRemainder). It may be a little unnatural, but I preferred the consistency of the definition.

Comment (awainverse) [reply Mathlib/Data/Nat/ModEq.lean:575]: If that's how the library already is, it seems fine to me.

**Review by fpvandoorn — APPROVED** [review]
Sorry that this took so long to review.
I have a minor suggestion about the module doc, but everything else looks good now.
After changing the module doc, you can merge this PR by typing `bors merge`.

bors d+

**Comment by hmonroe** [comment]
+r

**Comment by hmonroe** [comment]
bors merge

File: Mathlib/Data/Nat/ModEq.lean [hunk 16]
```diff
@@ -8,6 +8,9 @@ import Mathlib.Data.Int.Order.Lemmas
 import Mathlib.Logic.Basic
 import Mathlib.Tactic.NormNum
 import Mathlib.Tactic.GCongr.Core
+import Mathlib.Data.Nat.GCD.BigOperators
+import Mathlib.Data.List.MinMax
+import Mathlib.Data.List.FinRange
```
Comment (fpvandoorn) [inline Mathlib/Data/Nat/ModEq.lean:13]: You are adding a bunch of imports to a low-level file, which is not good. Please move the results that are about List/Multiset/Finset to a new file, e.g. `Data/Nat/ChineseRemainder`.

bors r-

Comment (fpvandoorn) [reply Mathlib/Data/Nat/ModEq.lean:13]: bors d-

File: Mathlib/Logic/Godel/GodelBetaFunction.lean [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+    (by simp[coprimes])
+    (by simpa using Set.pairwise_univ.mpr (pairwise_coprime_coprimes _)) : ℕ).pair
+  (supOfSeq l.get)!
+
+/-- **Gödel's Beta Function Lemma** -/
+lemma beta_function_lemma (l : List ℕ) (i : Fin l.length) :
+    beta (unbeta l) i = l.get i := by
+  simpa[beta, unbeta, coprimes] using mod_eq_of_modEq
+    ((chineseRemainderOfFinset l.get (coprimes l.get) Finset.univ
+      (by simp[coprimes])
```
Comment (eric-wieser) [inline Mathlib/Logic/Godel/GodelBetaFunction.lean:102]:
```suggestion
      (by simp [coprimes])
```

File: Mathlib/Logic/Godel/GodelBetaFunction.lean [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+    have : p = 1 := by
+      simpa[Nat.add_sub_cancel_left] using Nat.dvd_sub (le_add_right _ _) hn
+        (Dvd.dvd.mul_left this n)
+    simp[this] at pp
+    apply Nat.not_prime_one at pp
+    exact pp)
+
+variable {m : ℕ}
+
+private def supOfSeq (a : Fin m → ℕ) := max m (Finset.sup .univ a) + 1
```
Comment (eric-wieser) [inline Mathlib/Logic/Godel/GodelBetaFunction.lean:63]:
```suggestion
private def supOfSeq (a : Fin m → ℕ) : ℕ := max m (Finset.sup .univ a) + 1
```
I assume

File: Mathlib/Logic/Godel/GodelBetaFunction.lean [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+/-- Inverse of Gödel's Beta Function. This is similar to `Encodable.encodeList`, but it is easier
+to prove that it is arithmetically definable. -/
+def unbeta (l : List ℕ) : ℕ :=
+  (chineseRemainderOfFinset l.get (coprimes l.get) Finset.univ
+    (by simp[coprimes])
+    (by simpa using Set.pairwise_univ.mpr (pairwise_coprime_coprimes _)) : ℕ).pair
+  (supOfSeq l.get)!
+
+/-- **Gödel's Beta Function Lemma** -/
+lemma beta_function_lemma (l : List ℕ) (i : Fin l.length) :
```
[the remark this answers is not in the document]
Comment (hmonroe) [reply Mathlib/Logic/Godel/GodelBetaFunction.lean:99]: I renamed unbeta to beta_unbeta_coe--I hope that is what you meant

File: Mathlib/Data/Nat/ModEq.lean (hunk 16 above)
Comment (hmonroe) [reply Mathlib/Data/Nat/ModEq.lean]: Yes good point, I have done that

File: Mathlib/Logic/Godel/GodelBetaFunction.lean (hunk 19 above)
[the remark this answers is not in the document]
Comment (eric-wieser) [reply Mathlib/Logic/Godel/GodelBetaFunction.lean:100]: No, that's not what I meant. I meant this lemma should be renamed

bors r-

bors d-

File: Mathlib/Logic/Godel/GodelBetaFunction.lean [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+to prove that it is arithmetically definable. -/
+def beta_unbeta_coe (l : List ℕ) : ℕ :=
+  (chineseRemainderOfFinset l.get (coprimes l.get) Finset.univ
+    (by simp[coprimes])
+    (by simpa using Set.pairwise_univ.mpr (pairwise_coprime_coprimes _)) : ℕ).pair
+  (supOfSeq l.get)!
+
+/-- **Gödel's Beta Function Lemma** -/
+lemma beta_function_lemma (l : List ℕ) (i : Fin l.length) :
+    beta (beta_unbeta_coe l) i = l.get i := by
```
Comment (eric-wieser) [inline Mathlib/Logic/Godel/GodelBetaFunction.lean:101]: Just to be super clear:
```suggestion
def unbeta (l : List ℕ) : ℕ :=
  (chineseRemainderOfFinset l.get (coprimes l.get) Finset.univ
    (by simp[coprimes])
    (by simpa using Set.pairwise_univ.mpr (pairwise_coprime_coprimes _)) : ℕ).pair
  (supOfSeq l.get)!

/-- **Gödel's Beta Function Lemma** -/
lemma beta_unbeta_coe (l : List ℕ) (i : Fin l.length) :
    beta (unbeta l) i = l.get i := by
```

File: Mathlib/Logic/Godel/GodelBetaFunction.lean [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+  simpa[beta, beta_unbeta_coe, coprimes] using mod_eq_of_modEq
```
Comment (eric-wieser) [inline Mathlib/Logic/Godel/GodelBetaFunction.lean:102]:
```suggestion
  simpa [beta, beta_unbeta_coe, coprimes] using mod_eq_of_modEq
```
spaces before `[`

File: Mathlib/Data/Nat/ModEq.lean [hunk 22]
```diff
@@ -532,4 +541,5 @@ theorem odd_mod_four_iff {n : ℕ} : n % 2 = 1 ↔ n % 4 = 1 ∨ n % 4 = 3 :=
     fun h => Or.elim h odd_of_mod_four_eq_one odd_of_mod_four_eq_three⟩
 #align nat.odd_mod_four_iff Nat.odd_mod_four_iff
 
-end Nat
+lemma mod_eq_of_modEq {a b n} (h : a ≡ b [MOD n]) (hb : b < n) : a % n = b := by
+  simpa [show a % n = b % n from h] using mod_eq_of_lt hb
```
Comment (eric-wieser) [inline Mathlib/Data/Nat/ModEq.lean:545]: I assume this is `h.trans (mod_eq_of_lt hb)`?

Comment (hmonroe) [reply Mathlib/Data/Nat/ModEq.lean]: I changed to `Eq.trans h (mod_eq_of_lt hb)`
